### PR TITLE
Issue/427

### PR DESCRIPTION
**Pull request by ismaelbej**
This is an implementation of RFC 6637 to partially fix issue #427.

A few remarks of the code:
- Uses [Elliptic](https://github.com/indutny/elliptic/) for curve cryptography, but should be possible to add others backends like webcrypto or node's crypto.
- It uses [AES library](https://github.com/cryptocoinjs/aes)/ to implement key wrapping. I've failed to find a suitable AES decrypt function for RFC 3394 in the current code. But should be easy to replace.

It was tested against GnuPG version v2.1.8 + libgcrypt v1.7.0-beta262, some other details  are in the pull request [README-ECC.md](https://github.com/Jaxx-io/openpgpjs/blob/issue/427/README-ECC.md).

### CODE REVIEW

File: extra/bitcoin_interop.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+var openpgp = require('openpgp');
```
Comment (evilaliv3) [inline extra/bitcoin_interop.js]: This whole bitcoin example is nice but i would keep it out of the patch. We will document it in a different way.

Let's evaluate the removal before finishing the integration of the patch

**Comment by ismaelbej** [comment]
Removed 'const' from unit test and dropped bitcoin example.

File: src/packet/public_key.js [hunk 2]
```diff
@@ -100,8 +102,11 @@ PublicKey.prototype.read = function (bytes) {
     var p = 0;
 
     for (var i = 0; i < mpicount && p < bmpi.length; i++) {
-
-      this.mpi[i] = new type_mpi();
+      if (this.algorithm === 'ecdsa' && i === 0) {
```
Comment (ismaelbej) [inline src/packet/public_key.js]: Here is the issue related to the assumption that key packets only contain MPI. Also for 'ecdh' I've to add a similar condition.

File: src/packet/public_key.js [hunk 3]
```diff
@@ -100,8 +104,13 @@ PublicKey.prototype.read = function (bytes) {
     var p = 0;
 
     for (var i = 0; i < mpicount && p < bmpi.length; i++) {
-
-      this.mpi[i] = new type_mpi();
+      if ((this.algorithm === 'ecdsa' || this.algorithm === 'ecdh') && i === 0) {
```
Comment (ismaelbej) [inline src/packet/public_key.js]: Here is the issue related to the assumption that key packets only contain MPI. In order to support Ed25519 also a similar change have to be made.

File: src/openpgp.js [hunk 4]
```diff
@@ -374,6 +374,12 @@ function checkCleartextMessage(message) {
  * Format user ids for internal use.
  */
 function formatUserIds(options) {
+  if (!options.curve) {
```
Comment (ismaelbej) [inline src/openpgp.js]: This is a work around because some tests do not expect extra parameters like 'curve' and 'material'.

File: src/key.js [hunk 5]
```diff
@@ -968,13 +973,25 @@ export function generate(options) {
...
+    }
+    var material;
+    if (typeof(options.material) !== "undefined") {
+      material = options.material.subkey;
+    }
+    secretSubkeyPacket.algorithm = enums.read(enums.publicKey, subkeyType);
+    return secretSubkeyPacket.generate(options.numBits, options.curve, material);
   }
 
   function wrapKeyObject() {
```
Comment (ismaelbej) [inline src/key.js]: Here wrapKeyObject has to be modified to remove hash/encryption algorithms that should not be used with a given curve. For example a p521 key should not be used with sha1, but with sha256. Also prefer_hash_algorithm need to reflect that also.

**Comment by ismaelbej** [comment]
Replying to @sanjanarajan who asked why did I choose an external AES library. In order to support ECDH We've to implement key wrapping algorithm from RFC 3394 which uses AES encrypt and AES decrypt functions.

At the time the repository didn't have AES decrypt only AES encrypt was present (https://github.com/Jaxx-io/openpgpjs/blob/issue/427/src/crypto/cipher/aes.js), so we use the fast and light implementation from https://github.com/cryptocoinjs/aes. Before sending our pull request v2.0 was released with the new addition of asmcrypt-lite I did try to use AES decrypt from there but was unsuccessful. So we send our pull request with the annotation that the additional dependency can be removed if AES decrypt was made available to use.

**Comment by evilaliv3** [comment]
@sanjanarajan i'm noticing that you are try to move forward the ECC implementation.

that is great!  i'm interested to help retesting this and eventually study with you the possible adoption of WebCrypto as backend. are you already evaluating this possible backend alternative in your current implementation?

**Comment by sanjanarajan** [comment]
@evilaliv3 Yes, my current plan is to use a WebCrypto backend for the curves for which it is available, fall back to a Node Crypto backend when available, and use the elliptic backend as the last option. I have also done some refactoring to address the structure/style issues detailed in the previous comments. I am hoping to make a new pull request in the next week, looking forward to your feedback!

**Comment by bartbutler** [comment]
Closing as another version of this PR has just been merged!

**Pull request closed without merging**